libminifi/include/utils/tls/CertificateUtils.h

```diff
@@ -0,0 +1,61 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#ifdef OPENSSL_SUPPORT
+
+#include <openssl/ssl.h>
+
+#ifdef WIN32
+#include <windows.h>
+#include <wincrypt.h>
+#endif  // WIN32
+
+#include <memory>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+namespace tls {
+
+struct EVP_PKEY_deleter {
+  void operator()(EVP_PKEY* pkey) const { EVP_PKEY_free(pkey); }
+};
+using EVP_PKEY_unique_ptr = std::unique_ptr<EVP_PKEY, EVP_PKEY_deleter>;
+
+struct X509_deleter {
+  void operator()(X509* cert) const { X509_free(cert); }
```

Comment:
This is great to have as an utility class! Can we replace all usages of X509 free with this deleter? There are still usages in `SSLContextService.cpp` and `SourceInitiatedSubscriptionListener.cpp`.

Reply:
I have replaced `X509_free()` with `X509_unique_ptr` in `SourceInitiatedSubscriptionListener.cpp` (although this is part of the openwsman extension, which is disabled by default and does not compile at the moment -- I will fix that in a later PR.)

In `SSLContextService::addP12CertificateToSSLContext()`, I would have to replace
```c++
  const auto certs_guard = gsl::finally([pkey, cert, ca]() {
    EVP_PKEY_free(pkey);
    X509_free(cert);
    sk_X509_pop_free(ca, X509_free);
  });
```
with
```c++
  utils::tls::EVP_PKEY_unique_ptr pkey_ptr{pkey};
  utils::tls::X509_unique_ptr cert_ptr{cert};
  const auto ca_deleter = gsl::finally([ca] { sk_X509_pop_free(ca, X509_free); });
```
ie., introduce 3 deleter variables instead of 1.  Do you think that is worth it?

Reply:
we could wrap the `PKCS12_parse` to return an optional structure containing these smart pointers, unclear if worth the added complexity

Reply:
Yeah, if `PKCS12_parse()` was used anywhere else, I would do that, but for this one usage I don't think it's worth it.

Reply:
>  Do you think that is worth it?

Yes, I think your change clears it up a lot.

Reply:
done: https://github.com/apache/nifi-minifi-cpp/pull/947/commits/6e90811fd2e8f4bf261d519e82dc6d87113fddeb